### PR TITLE
Cache backend

### LINKED ISSUE #21: Fix the Cache backend for sliced image
We want to be able to check if the data has been downloaded locally before we load the imagestack. If it exists, we want to use the local copy in lieu of a web link.

### PR DESCRIPTION
**Pull request by shanaxel42**
Added caching backend.

Removed `test_numpy_requires_seekable` from `test_http_backend.py` because CachingBackend always reads from the cache to return, and it always returns a seekable handle so the error will never be thrown.

Fixes #21

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,3 +4,5 @@ numpy
 requests
 scikit-image
 six
+diskcache
+
```
Comment (ttung) [inline requirements.txt]: remove extra NL.
keep list sorted.

File: slicedimage/backends/_caching.py [hunk 2]
```diff
@@ -1,6 +1,9 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
-import os
+import io
+
+from diskcache import Cache
+from io import BytesIO
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: this import should be grouped with stdlib imports.  we're following the PEP8 convention (more specifically the smarkets variant described [here](https://github.com/PyCQA/flake8-import-order#styles))

generally, i also avoid importing the same module twice, so I would just use `io.BytesIO` in this module.

File: slicedimage/backends/_caching.py [hunk 3]
```diff
@@ -9,23 +12,23 @@ class CachingBackend(Backend):
     def __init__(self, cacheroot, authoritative_backend):
         self._cacheroot = cacheroot
         self._authoritative_backend = authoritative_backend
+        self.cache = Cache(cacheroot, size_limit=int(5e9))
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: consider making 5e9 a constant in this module.

alternatively, you can do:

```
    def __init__(self, cacheroot, authoritative_backend, size_limit=int(5e9)):
        self._cacheroot = cacheroot
        self._authoritative_backend = authoritative_backend
        self.cache = Cache(cacheroot, size_limit=size_limit)
```

rationale: easier to find the constant.

File: slicedimage/backends/_caching.py [hunk 4]
```diff
@@ -9,23 +12,23 @@ class CachingBackend(Backend):
...
-                    self._authoritative_backend.read_file_handle(name, checksum_sha1) as sfh:
-                while True:
-                    data = sfh.read(128 * 1024)
-                    if len(data) == 0:
-                        break
-                    dfh.write(data)
-
-        return lambda: open(cachepath, "r")
+        def returned_callable():
+            if checksum_sha1:
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: I would rename all these variables `checksum_sha256` so it's not misleading.

File: slicedimage/backends/_caching.py [hunk 5]
```diff
@@ -9,23 +12,23 @@ class CachingBackend(Backend):
...
+                if not self.cache.get(checksum_sha1):
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: does this actually retrieve the cached item?  if so, you're reading the data twice.

File: slicedimage/backends/_caching.py [hunk 6]
```diff
@@ -9,23 +12,23 @@ class CachingBackend(Backend):
...
+                    sfh = self._authoritative_backend.read_file_handle(name)
+                    self.cache.set(checksum_sha1, sfh.data)
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: I would do sfh.read() because the contract is that it's a file handle.  That it provides the data in .data is an implementation artifact.

File: slicedimage/backends/_caching.py [hunk 7]
```diff
@@ -9,23 +12,23 @@ class CachingBackend(Backend):
...
+                file_data = self.cache.read(checksum_sha1)
+                # The DiskCache library returns the cache data as bytes instead of a buffered reader
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: proposal:

`If the data is small enough, the DiskCache library returns the cache data as bytes instead of a buffered reader.  In that case, we want to wrap it in a file-like object.`

File: slicedimage/io.py [hunk 8]
```diff
@@ -10,7 +10,7 @@
 from six.moves import urllib
 
 from slicedimage.urlpath import pathjoin, pathsplit
-from .backends import DiskBackend, HttpBackend
+from .backends import DiskBackend, HttpBackend, CachingBackend
```
Comment (ttung) [inline slicedimage/io.py]: sort pls.

File: slicedimage/io.py [hunk 9]
```diff
@@ -26,17 +26,14 @@ def infer_backend(baseurl, allow_caching=True):
 
     if parsed.scheme in ("http", "https"):
         backend = HttpBackend(baseurl)
+        if allow_caching:
+            backend = CachingBackend("~/tmp", backend)
```
Comment (ttung) [inline slicedimage/io.py]: As far as I know, this does not do path expansion.  I think you want:

`backend = CachingBackend(os.path.expanduser("~/.starfish-cache"), backend)`

File: tests/io/v0_0_0/test_caching_backend.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+                if time.time() > end:
+                    raise
+
+    def tearDown(self):
+        for context in self.contexts:
+            context.__exit__(*sys.exc_info())
+
+    def test_cached_backend(self):
+        """
+        Generate a tileset consisting of a single TIFF tile.  Deposit it where the HTTP server can
```
Comment (ttung) [inline tests/io/v0_0_0/test_caching_backend.py]: This docblock should reflect your testing strategy.

File: tests/io/v0_0_0/test_caching_backend.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        find the tileset, and fetch it.
+        """
+        # write the tiff file
+        data = np.random.randint(0, 65535, size=(100, 100), dtype=np.uint16)
```
Comment (ttung) [inline tests/io/v0_0_0/test_caching_backend.py]: This looks good; consider also testing a larger file since we know that it behaves slightly differently.

File: tests/io/v0_0_0/test_caching_backend.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                        0.0001,
+                    ]
+                },
+                "indices": {
+                    "hyb": 0,
+                    "ch": 0,
+                },
+                "file": "tile.tiff",
+                "format": "tiff",
+                "sha256": random.getrandbits(128)
```
Comment (ttung) [inline tests/io/v0_0_0/test_caching_backend.py]: We should provide a real checksum here.

File: tests/io/v0_0_0/test_caching_backend.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        # Now delete tile.tff from the http server and make sure Reader.parse_doc still works
+        os.remove(os.path.join(self.tempdir.name, "tile.tiff"))
+        result = slicedimage.Reader.parse_doc(
+            "tileset.json",
+            "http://localhost:{port}/".format(port=self.port))
+
+        self.assertTrue(np.array_equal(list(result.tiles())[0].numpy_array, data))
+
+
+def _unused_tcp_port():
```
Comment (ttung) [inline tests/io/v0_0_0/test_caching_backend.py]: This is the same as in `test_http_backend.py`.

Ideal: precede this PR with one that moves this common stuff to `tests/utils`, then stack this change on top of that.  **OR**  follow this PR with another PR  that deduplicates the common stuff.
Less ideal: move this common stuff to `tests/utils` in this PR.

File: tests/io/v0_0_0/test_http_backend.py [hunk 14]
```diff
@@ -85,6 +86,8 @@ def test_tiff(self):
                 },
                 "file": "tile.tiff",
                 "format": "tiff",
+                "sha256": random.getrandbits(128)
```
Comment (ttung) [inline tests/io/v0_0_0/test_http_backend.py]: Is this change desired?

File: tests/io/v0_0_0/test_http_backend.py [hunk 15]
```diff
@@ -132,48 +135,6 @@ def test_numpy(self):
 
         self.assertTrue(np.array_equal(list(result.tiles())[0].numpy_array, tile.numpy_array))
 
-    def test_numpy_requires_seekable(self):
```
Comment (ttung) [inline tests/io/v0_0_0/test_http_backend.py]: In your PR description, you should explain why this was removed (basically what you explained to me in person is reasonable).

**Comment by ttung** [comment]
Also, if you write "Fixes #issue-number" in your PR description, then it'll close the github issue when you merge.  So satisfying!

File: tests/io/v0_0_0/test_http_backend.py (hunk 15 above)
Comment (shanaxel42) [inline tests/io/v0_0_0/test_http_backend.py]: If caching is allowed (which it is in this test), CachingBackend is used. CachingBackend always reads from the cache to return, and it always returns in a readable format so this test was no longer necessary, the error will never be thrown.

File: slicedimage/backends/_caching.py [hunk 16]
```diff
@@ -1,31 +1,37 @@
...
+import io
+
+from diskcache import Cache
 
 from ._base import Backend
 
+size_limit = 5e9
+
 
 class CachingBackend(Backend):
```
Comment (dganguli) [inline slicedimage/backends/_caching.py]: Consider re-naming to "CacheRulesEverythingAroundMeBackend" or C.R.E.A.M.Backend? https://www.youtube.com/watch?v=PBwAxmrE194

File: tests/io/v0_0_0/test_caching_backend.py [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+import slicedimage
+
+pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))  # noqa
+sys.path.insert(0, pkg_root)  # noqa
+
+from tests.utils import TemporaryDirectory
+
+
+class TestCachingBackend(unittest.TestCase):
```
Comment (dganguli) [inline tests/io/v0_0_0/test_caching_backend.py]: this is awesome. i don't understand a lot of the code, but if this test passes, then i'm extremely happy.

File: slicedimage/backends/_caching.py (hunk 16 above)
Comment (shanaxel42) [reply slicedimage/backends/_caching.py]: Best comment.

File: slicedimage/backends/_caching.py [hunk 18]
```diff
@@ -1,31 +1,37 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
-import os
+import io
+
+from diskcache import Cache
 
 from ._base import Backend
 
+size_limit = 5e9
```
Comment (ttung) [inline slicedimage/backends/_caching.py]: There aren't really true constants in python, but "constants" in python are generally given all caps names, i.e., SIZE_LIMIT

File: tests/io/v0_0_0/test_caching_backend.py (hunk 13 above)
Comment (ttung) [reply tests/io/v0_0_0/test_caching_backend.py]: So if you rebase on top of master, which contains #32, you should be able to remove all this extra code!

File: tests/io/v0_0_0/test_http_backend.py [hunk 19]
```diff
@@ -83,7 +83,7 @@ def test_tiff(self):
                     "ch": 0,
                 },
                 "file": "tile.tiff",
-                "format": "tiff",
```
Comment (ttung) [inline tests/io/v0_0_0/test_http_backend.py]: please revert this change.

Rationale for the trailing commas: if something is added, then we don't affect the line above.

File: tests/io/v0_0_0/test_http_backend.py (hunk 15 above)
Comment (ttung) [reply tests/io/v0_0_0/test_http_backend.py]: Thanks!  When you merge, github might be dumb and not slurp in this text as part of your commit description.  You may have to copy and paste it over.

File: slicedimage/backends/_caching.py [hunk 20]
```diff
@@ -1,31 +1,37 @@
...
-            with open(cachepath, "w") as dfh, \
-                    self._authoritative_backend.read_file_handle(name, checksum_sha1) as sfh:
-                while True:
-                    data = sfh.read(128 * 1024)
-                    if len(data) == 0:
-                        break
-                    dfh.write(data)
-
-        return lambda: open(cachepath, "r")
+    def read_file_handle_callable(self, name, checksum_sha256=None, seekable=False):
```
Comment (ambrosejcarr) [inline slicedimage/backends/_caching.py]: I know that most of the code here doesn't have docstrings, but now is as good a time as any to start :)

File: tests/io/v0_0_0/test_caching_backend.py [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    def setUp(self, timeout_seconds=5):
```
Comment (ambrosejcarr) [inline tests/io/v0_0_0/test_caching_backend.py]: Please add a short description of what this setup does. :-)

File: tests/io/v0_0_0/test_caching_backend.py (hunk 10 above)
Comment (ambrosejcarr) [reply tests/io/v0_0_0/test_caching_backend.py]: The description of the strategy seems like it takes for granted the fact that the backend will pull from the cache. Is there a way to explicitly test that it is doing so? 

i.e. if for some reason it just downloaded the doc from the URL twice, you'd expect this test to pass.

File: tests/io/v0_0_0/test_caching_backend.py [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+def _unused_tcp_port():
+    """
+    Return an unused TCP port.
+    """
+    with contextlib.closing(socket.socket()) as sock:
+        sock.bind(('127.0.0.1', 0))
+        return sock.getsockname()[1]
+
+
+class ContextualChildProcess(object):
```
Comment (ambrosejcarr) [inline tests/io/v0_0_0/test_caching_backend.py]: For this and other objects you define, the python standard is to define objects above/before the methods that call them. Please move these up.

**Comment by ambrosejcarr** [comment]
> final review changes e00a3ab
heh heh heh. 👿

File: tests/io/v0_0_0/test_caching_backend.py (hunk 10 above)
Comment (shanaxel42) [reply tests/io/v0_0_0/test_caching_backend.py]: Well I'm deleting the .tiff file off the http server, so the second time I get the result would error if it tried to use the HttpBackend (with a file not found) instead of the CacheBackend. I updated the doctoring to explain that better though

Comment (ambrosejcarr) [reply tests/io/v0_0_0/test_caching_backend.py]: Oh! Just add "from the server" to the text block. :-)

If it is there, then sorry for the dumb question.
